### PR TITLE
HDDS-12484. Allow ini files to configure ColumnFamilyOptions for rocksDB

### PR DESCRIPTION
**Pull request by Galsza**
## What changes were proposed in this pull request?
https://issues.apache.org/jira/browse/HDDS-12484
Ozone only had a way to change DBOptions from a ini configuration file before, now it's possible to also specify ColumnFamilyOptions.

This PR gives the option to fully customize all RocksDB options. It also enables the user to specify a path from which the configuration file will be read and used.

There are a few known issues regarding this: 7.x versions of RocksDB have a bug where TableFormatConfigs are not read from a file. This is fixed in later versions, but upgrading Ozone to a newer version requires a bigger effort, see here: https://issues.apache.org/jira/browse/HDDS-12392 . For handling this the Ozone default configs are set, and I've created the following Jira to change these places once 9.x versions of RocksDB are used: https://issues.apache.org/jira/browse/HDDS-12695

Due to RocksDB calls invoking native libraries the resource management has to be carefully observed and some unusual behaviour is created in java. Mainly freeing up resources in a catch block.

Snapshots shouldn't be affected as they should remain unchanged even if the user changes cf settings.

## How was this patch tested?

Manually firing up a cluster through docker and checking if the cfoptions really did change, and added unit tests.

### CODE REVIEW

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBStoreBuilder.java [hunk 1]
```diff
@@ -142,10 +134,6 @@ private DBStoreBuilder(ConfigurationSource configuration,
     this.rocksDbStat = configuration.getTrimmed(
         OZONE_METADATA_STORE_ROCKSDB_STATISTICS,
         OZONE_METADATA_STORE_ROCKSDB_STATISTICS_DEFAULT);
-    this.rocksDbCfWriteBufferSize = (long) configuration.getStorageSize(
```
Comment (chungen0126) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBStoreBuilder.java:145]: Since OZONE_METADATA_STORE_ROCKSDB_CF_WRITE_BUFFER_SIZE is no longer used, the TestOMSnapshotDAG should also be rewritten. It is still using it.
In addition, I think we need a default ini file to set `write_buffer_size` as the default size in Ozone is different from rocksdb.

**Review by chungen0126 — COMMENTED** [review]
Thanks @Galsza for the patch. Left some comments.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/db/DatanodeDBProfile.java [hunk 2]
```diff
@@ -115,34 +114,28 @@ private StorageBasedProfile(DBProfile profile) {
...
+        ConfigurationSource config, Path pathToDb, String cfName) {
       final MemoizedSupplier<ManagedColumnFamilyOptions> supplier =
-          MemoizedSupplier.valueOf(() -> createColumnFamilyOptions(config));
+          MemoizedSupplier.valueOf(() -> {
+            ManagedColumnFamilyOptions options =
+                baseProfile.getColumnFamilyOptions(pathToDb, cfName);
+            options.setReused(true);
+            return options.closeAndSetTableFormatConfig(
+                getBlockBasedTableConfig(pathToDb, cfName, config));
+          });
```
[the remark this answers is not in the document]
Comment (Galsza) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/db/DatanodeDBProfile.java]: Thanks for the suggestion it's indeed clearer.

File: hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedBlockBasedTableConfig.java [hunk 3]
```diff
@@ -69,4 +69,37 @@ public void close() {
       }
     }
   }
+
+  public synchronized ManagedBlockBasedTableConfig setAllProperties(
```
[the remark this answers is not in the document]
Comment (Galsza) [reply hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedBlockBasedTableConfig.java]: I think once RocksDB will be updated this code needs to be revisited. Currently it's unused so I deleted it.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBConfigFromFile.java [hunk 4]
```diff
@@ -106,38 +110,86 @@ public static String getOptionsFileNameFromDB(String dbFileName) {
...
+    if (!generatedDBPath.toFile().exists()) {
+      LOG.warn("Error trying to read generated rocksDB file: {}, file does not exists.", generatedDBPath);
+      return null;
+    }
+    ManagedDBOptions options = new ManagedDBOptions();
+    try (ManagedConfigOptions configOptions = new ManagedConfigOptions()) {
+      OptionsUtil.loadOptionsFromFile(configOptions, generatedDBPath.toString(), options, descriptors);
+    } catch (RocksDBException rdEx) {
+      options.close();
+      closeDescriptors(descriptors);
```
[the remark this answers is not in the document]
Comment (Galsza) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBConfigFromFile.java]: Thanks for the suggestion, I changed it

File: hadoop-hdds/common/src/main/resources/ozone-default.xml [hunk 5]
```diff
@@ -2680,6 +2680,14 @@
       Absolute path to HDDS metadata dir.
     </description>
   </property>
+  <property>
+    <name>rocks.db.config.path</name>
+    <value/>
+    <tag>OZONE, OM</tag>
```
[the remark this answers is not in the document]
Comment (Galsza) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml:2686]: Yes and I've also realized that there are places where we are incorrectly using the config options, so this needs to be revisited in an upcoming commit soon.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBStoreBuilder.java (hunk 1 above)
Comment (Galsza) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBStoreBuilder.java:145]: Most other configs are few enough that they are hardcoded programatically into Ozone, I'm not sure we want to change that here for the default ini. However having an example ini would be more failsafe when the user gets control.

I'm not sure how the default ini should be handled for example in test environments. If we rely on the default ini being always there then I'm not sure how to simply do it for all test environments. Do you have a recommendation? In the case we have to modify most docker-yaml files I'd prefer to defer that change to a separate Jira.

File: hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/utils/db/TestDBConfigFromFile.java [hunk 6]
```diff
@@ -90,23 +70,21 @@ public void readFromFile() throws IOException {
...
-
-    final DBOptions options = DBConfigFromFile.readFromFile("badfile.db.ini",
-        columnFamilyDescriptors);
-
+    final DBOptions options = DBConfigFromFile.readDBOptionsFromFile(Paths.get("badfile.db.ini"));
     // This has to return a Null, since we have config defined for badfile.db
     assertNull(options);
   }
+
+  @Test
```
Comment (jojochuang) [inline hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/utils/db/TestDBConfigFromFile.java:78]: Please also check that edge cases (such as an empty or misconfigured ini file) are covered in the test suite.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/DatanodeSchemaThreeDBDefinition.java [hunk 7]
```diff
@@ -110,17 +113,14 @@ public DatanodeSchemaThreeDBDefinition(String dbPath,
     DatanodeDBProfile dbProfile = DatanodeDBProfile
         .getProfile(config.getEnum(HDDS_DB_PROFILE, HDDS_DEFAULT_DB_PROFILE));
 
-    ManagedColumnFamilyOptions cfOptions =
-        dbProfile.getColumnFamilyOptions(config);
-    // Use prefix seek to mitigating seek overhead.
-    // See: https://github.com/facebook/rocksdb/wiki/Prefix-Seek
```
Comment (jojochuang) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/DatanodeSchemaThreeDBDefinition.java:116]: Please add these comments in getCFOptions()

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBProfile.java [hunk 8]
```diff
@@ -44,51 +49,98 @@ public String toString() {
...
+            .setCreateMissingColumnFamilies(createMissingColumnFamilies);
+        return dbOptions;
+      }
+      return option;
+    }
+
+    @Override
+    public ManagedColumnFamilyOptions getColumnFamilyOptions(Path dbPath, String cfName) {
+      ManagedColumnFamilyOptions option = null;
+      if (Objects.isNull(dbPath)) {
```
Comment (jojochuang) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBProfile.java:85]: The default dbPath is empty (not null). What happens when it's empty?

Comment (Galsza) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBProfile.java:85]: The default value for this configuration is empty. In that case it tries first reading that as a file, which will obviously fail, and then does a fallback which tries the same functionality as before this pull request. So if it's a simple db name it tries searching for it on the env variable provided or in the OZONE_CONFIG_DIR.

**Review by errose28 — COMMENTED** [review]
Thanks for working on this @Galsza. Are we planning to get rid of the `DBProfile` concept entirely, either in this PR or a follow up? It might be easier to just ship different example INI files for various configurations than maintain this extra layer in the code.

**Comment by Galsza** [comment]
@errose28 I'm not sure when I can prioritize removing the DBProfile concept, but it should be done some time. The code in its current form is hard to maintain.

**Comment by errose28** [comment]
Noted, I was just wondering what the roadmap was. If it is out of scope for this change, can you file a follow-up jira under the same epic so it does not get lost?

File: pom.xml [hunk 9]
```diff
@@ -1559,6 +1559,7 @@
                       <allowedImport>org.rocksdb.TransactionLogIterator.BatchResult</allowedImport>
                       <allowedImport>org.rocksdb.TickerType</allowedImport>
                       <allowedImport>org.rocksdb.LiveFileMetaData</allowedImport>
+                      <allowedImport>org.rocksdb.BlockBasedTableConfig</allowedImport>
```
Comment (fapifta) [inline pom.xml:1562]: Do we need this for some reason? If not, then good, if there is a place we use the BlockBasedTableConfig type outside the managed-rocksdb module, then we need to avoid that and use the ManagedBlockBasedTableConfig instead outside the managed rocksdb module.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/DatanodeSchemaThreeDBDefinition.java [hunk 10]
```diff
@@ -110,17 +113,14 @@ public DatanodeSchemaThreeDBDefinition(String dbPath,
...
-    cfOptions.useFixedLengthPrefixExtractor(getContainerKeyPrefixLength());
+    Path pathToDb = Paths.get(
+        config.get(HddsConfigKeys.DATANODE_DB_CONFIG_PATH, HddsConfigKeys.DATANODE_DB_CONFIG_PATH_DEFAULT));
 
-    BLOCK_DATA.setCfOptions(cfOptions);
-    METADATA.setCfOptions(cfOptions);
-    DELETE_TRANSACTION.setCfOptions(cfOptions);
-    FINALIZE_BLOCKS.setCfOptions(cfOptions);
-    LAST_CHUNK_INFO.setCfOptions(cfOptions);
+    BLOCK_DATA.setCfOptions(getCFOptions(config, dbProfile, pathToDb, "block_data"));
```
Comment (peterxcli) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/DatanodeSchemaThreeDBDefinition.java:119]: How about sharing those column families name as constant?
```java
final string BLOCK_DATA_TABLE = "block_data";
...
public static final DBColumnFamilyDefinition<String, BlockData>
      BLOCK_DATA =
      new DBColumnFamilyDefinition<>(
          BLOCK_DATA_TABLE,
          FixedLengthStringCodec.get(),
          BlockData.getCodec());
...
BLOCK_DATA.setCfOptions(getCFOptions(config, dbProfile, pathToDb, BLOCK_DATA_TABLE));
```

File: hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedColumnFamilyOptions.java [hunk 11]
```diff
@@ -49,7 +50,9 @@ public synchronized ManagedColumnFamilyOptions setTableFormatConfig(
       if (!((ManagedBlockBasedTableConfig) previous).isClosed()) {
         throw new IllegalStateException("Overriding an unclosed value.");
       }
-    } else if (previous != null) {
+    } else if (!(previous instanceof BlockBasedTableConfig) && previous != null) {
```
Comment (peterxcli) [inline hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedColumnFamilyOptions.java:53]:
```suggestion
    } else if (previous != null && !(previous instanceof BlockBasedTableConfig)) {
```

**Comment by Galsza** [comment]
@peterxcli , @jojochuang could you take another look on this pull request?

**Comment by adoroszlai** [comment]
@Galsza please fix pmd failure

```
PMD Failure: org.apache.hadoop.hdds.utils.db.DBProfile:203 Rule:FieldDeclarationsShouldBeAtStartOfClass Priority:3 Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes..
```

https://github.com/Galsza/ozone/actions/runs/14737271726/job/41366446259#step:16:17

**Comment by adoroszlai** [comment]
Thanks @Galsza for updating the patch.

@jojochuang @peterxcli Please take another look.  If you are OK with the current changes, please trigger CI, too.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBProfile.java [hunk 12]
```diff
@@ -44,51 +48,98 @@ public String toString() {
...
+        dbOptions
+            .setIncreaseParallelism(Runtime.getRuntime().availableProcessors())
+            .setMaxBackgroundCompactions(maxBackgroundCompactions)
+            .setMaxBackgroundFlushes(maxBackgroundFlushes)
+            .setBytesPerSync(bytesPerSync)
+            .setCreateIfMissing(createIfMissing)
+            .setCreateMissingColumnFamilies(createMissingColumnFamilies);
+        return dbOptions;
+      }
+      return option;
```
Comment (peterxcli) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBProfile.java:78]: just a cleanup
```suggestion
      if (option != null) {
        LOG.info("Using RocksDB DBOptions from {}.ini file", optionsPath);
        return option;
      } 
      final int maxBackgroundCompactions = 4;
      final int maxBackgroundFlushes = 2;
      final long bytesPerSync = toLong(StorageUnit.MB.toBytes(1.00));
      final boolean createIfMissing = true;
      final boolean createMissingColumnFamilies = true;
      option = new ManagedDBOptions();
      option
          .setIncreaseParallelism(Runtime.getRuntime().availableProcessors())
          .setMaxBackgroundCompactions(maxBackgroundCompactions)
          .setMaxBackgroundFlushes(maxBackgroundFlushes)
          .setBytesPerSync(bytesPerSync)
          .setCreateIfMissing(createIfMissing)
          .setCreateMissingColumnFamilies(createMissingColumnFamilies);
      return option;
```

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBStoreBuilder.java [hunk 13]
```diff
@@ -324,8 +314,7 @@ private Set<TableConfig> makeTableConfigs() {
 
       if (options == null) {
         LOG.debug("using default column family options for table: {}", name);
-        tableConfigs.add(new TableConfig(name,
-                getCfOptions(rocksDbCfWriteBufferSize)));
+        tableConfigs.add(new TableConfig(name, getCfOptions(name)));
```
Comment (peterxcli) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBStoreBuilder.java:317]: Would this be called after `defaultCfOptions` is set? If yes, the getCfOptions could be replaced by `defaultCfOptions` as the log here says: "using default column family options for table: xxx"

**Review by peterxcli — COMMENTED** [review]
Thanks @Galsza for updating the patch! Left more comments~

**Review by peterxcli — COMMENTED** [review]
Thanks @Galsza for updating the patch! Prod code looks good to me, but it seems there is no test to test the happy path of reading the CfOption from ini file. Have you tested it locally?

**Comment by Galsza** [comment]
@peterxcli Sorry for the late reply, yes I've retested the latest changes locally now. Both the fallback in case we can't find the missing file and the happy path is working as expected.

**Comment by Galsza** [comment]
@swamirishi Could you please review this PR? Your insights would be valuable here

**Comment by Galsza** [comment]
@swamirishi This is a fairly old pr now, but I updated it, could you please review it?

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBConfigFromFile.java [hunk 14]
```diff
@@ -103,37 +108,87 @@ public static String getOptionsFileNameFromDB(String dbFileName) {
...
+      if (descriptor != null) {
+        resultCfOptions = new ManagedColumnFamilyOptions(descriptor.getOptions());
       }
+    } finally {
+      closeDescriptors(descriptors);
+    }
+    return resultCfOptions;
+  }
+
+  private static ManagedDBOptions readFromFile(Path dbPath, List<ColumnFamilyDescriptor> descriptors)
```
Comment (swamirishi) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBConfigFromFile.java]: This function can return a AutoCloseable wrapper which would close the descriptors when close is called. Moreover we are always creating a new ArrayList<>() for descriptors so it would be better if just initialize descriptor inside this function

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBProfile.java [hunk 15]
```diff
@@ -44,51 +48,92 @@ public String toString() {
     }
 
     @Override
-    public ManagedColumnFamilyOptions getColumnFamilyOptions() {
-      // Write Buffer Size -- set to 128 MB
-      final long writeBufferSize = toLong(StorageUnit.MB.toBytes(128));
+    public ManagedDBOptions getDBOptions(Path optionsPath) {
```
Comment (swamirishi) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBProfile.java]: This is a bad idea to add the optionsPath in DB profile. Reading from ini file should be done on the DBStoreBuilder and not on the DBProfile. I believe we can load the ini file and not load from DB profile if the ini file exists on the DBStoreBuilder side. We would have to implement the same optionsPath read logic for each and every enum which is a bad idea

**Review by swamirishi — CHANGES_REQUESTED** [review]
@Galsza thanks for the patch I believe we can make the change of reading from ini file in DBStoreBuilder

**Comment by priyeshkaratha** [comment]
@fapifta @adoroszlai @jojochuang @errose28 @swamirishi 
Can anyone help me to merge this changes?

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBConfigFromFile.java [hunk 16]
```diff
@@ -103,37 +108,87 @@ public static String getOptionsFileNameFromDB(String dbFileName) {
...
+        throw new RocksDBException("There was an error opening rocksDB Options file: " + rdEx.getMessage());
+      }
+      return options;
+    } finally {
+      //readDBOptions will be freed once the store using it is closed, but the descriptors need to be closed.
+      closeDescriptors(descriptors);
+    }
+  }
+
+  public static ManagedColumnFamilyOptions readCFOptionsFromFile(Path optionsPath, String cfName)
```
Comment (swamirishi) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBConfigFromFile.java:137]: Both readCFOptionsFromFile and readDBOptionsFromFile has a very similar logic for initializing can we create a private static function which unifies the common parts in the function

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBConfigFromFile.java [hunk 17]
```diff
@@ -103,37 +108,87 @@ public static String getOptionsFileNameFromDB(String dbFileName) {
...
+    }
+    List<ColumnFamilyDescriptor> descriptors = new ArrayList<>();
+    String defaultColumnFamilyString = StringUtils.toEncodedString(DEFAULT_COLUMN_FAMILY, StandardCharsets.UTF_8);
+    String validatedCfName = StringUtils.isEmpty(cfName) ? defaultColumnFamilyString : cfName;
+    ManagedColumnFamilyOptions resultCfOptions = null;
+    try (ManagedConfigOptions ignored = new ManagedConfigOptions(); ManagedDBOptions options = new ManagedDBOptions()) {
+      OptionsUtil.loadOptionsFromFile(ignored, generatedDBPath.toString(), options, descriptors);
+      ColumnFamilyDescriptor descriptor = descriptors.stream()
+          .filter(desc -> StringUtils.toEncodedString(desc.getName(), StandardCharsets.UTF_8).equals(validatedCfName))
+          .findAny().orElse(null);
```
Comment (swamirishi) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/DBConfigFromFile.java:152]:
```suggestion
          .findFirst().orElse(null);
```

**Review by swamirishi — APPROVED** [review]
LGTM thanks for the patch @Galsza

**Comment by swamirishi** [comment]
@Galsza you can address the nitpicky review comments as part of a follow up patch. Thanks @jojochuang @peterxcli @errose28 @adoroszlai @chungen0126 for reviewing the patch

**Comment by adoroszlai** [comment]
@swamirishi please resolve Jira issue after merging PRs
